import os
import gensim


def read_corpus(pmids, abstracts):
    tagged_corpus = []
    for pmid, abstract in zip(pmids, abstracts):
        tokens = gensim.utils.simple_preprocess(abstract[0])
        tagged_corpus.append(gensim.models.doc2vec.TaggedDocument(tokens, [pmid]))
    return tagged_corpus


def train_model(tagged_corpus, output_filepath):
    model = gensim.models.doc2vec.Doc2Vec(vector_size=200, window=5, min_count=1, epochs=5)  # uses PV-DBOW architecture by default
    model.build_vocab(tagged_corpus)
    model.train(tagged_corpus, total_examples=model.corpus_count, epochs=model.epochs)
    model.save(output_filepath)

    # Generates vectors for each document of the training corpus
    for doc_id in range(len(tagged_corpus)):
        inferred_vector = model.infer_vector(tagged_corpus[doc_id].words)
        print(inferred_vector)

    # Testing the model
    # inferred_vector = model.infer_vector(test_corpus[0].words)
    # sims = model.dv.most_similar([inferred_vector], topn=len(model.dv))  # uses cosine similarity

    # Compare and print the most/median/least similar documents from the train corpus
    # print('Test Document ({}): «{}»\n'.format(0, ' '.join(test_corpus[0].words)))
    # print(u'SIMILAR/DISSIMILAR DOCS PER MODEL %s:\n' % model)
    # for label, index in [('MOST', 0), ('MEDIAN', len(sims)//2), ('LEAST', len(sims) - 1)]:
    #     print(u'%s %s: \n' % (label, sims[index])) #, ' '.join(train_corpus[sims[index][0]].words)))


if __name__ == "__main__":
    pmids = [9102202, 9105560, 9112375, 9112401, 9120013]
    abstracts = [['brca1 brca2 mutation confer increase risk development breast cancer, number additional, currently largely unknown, somatic genetic defect must also accumulate breast epithelial cell malignancy develops. evaluate nature additional somatic genetic defect, perform genome-wide survey comparative genomic hybridization breast cancer 21 brca1 mutation carrier, 15 brca2 mutation carrier, 55 unselected control. total number genetic change almost two time high tumor brca1 brca2 mutation carrier control group. brca1 tumor, loss 5q (86 %), 4q (81 %), 4p (64 %), 2q (40 %), 12q (40 %) significantly common control group (7-13 %). brca2 tumor characterize high frequency 13q (73 %) 6q (60 %) loss gain 17q22-q24 (87 %) 20q13 (60 %) compare prevalence change control group (12-18 %). conclusion, accumulation somatic genetic change tumor progression may follow unique pathway individual genetically predispose cancer, especially brca1 gene. activation loss gene affected chromosomal region may select tumor progression cell lack functional brca1 brca2. identification gene could provide target therapeutic intervention early diagnosis.'], ['familial ligand-defective apolipoprotein (apo) b-100 (fdb) autosomal codominant disorder may give rise hypercholesterolaemia. cause substitution glutamine arginine codon 3500 apo b gene (apo b r3500q), result decrease binding low density lipoprotein (ldl) ldl receptor. order search mutation region apo b gene, screen genomic dna, obtain 412 hypercholesterolaemic individual, use heteroduplex analysis. additional heteroduplex band observe follow analysis dna 11 individual, nine heterozygous apo b r3500q. two remain individual, celtic origin, show dna sequence heterozygous c--> transition nucleotide 10800 apo b gene, result substitution cysteine arginine codon 3531 (apo b r3531c). strong family history atherosclerosis family study reveal four individual heterozygous mutation, three hypercholesterolaemic. individual heterozygous apo b r3531c r3500q mean +/- s.e.m. cholesterol concentration 7.82 +/- 0.68 8.53 +/- 0.31 mmol/l, respectively. value significantly high value 5.51 +/- 0.23 mmol/l observe unaffected relative. finding suggest apo b r3531c less common uk give rise less severe form hypercholesterolaemia classical 3500 mutation. one family, r3531c mutation occur haplotype, compatible previously assign mutation north american family also celtic origin. consistent mutation inherit common distant ancestor individual celtic origin.'], ['study, partial hamster complementary dna encode ercc-1, member dna excision repair gene family, clone. nucleic acid amino acid sequence highly homologous human mouse ercc-1. hamster ercc-1 gene express 1.2-kilobase message culture chinese hamster ovary cell. northern (rna) blot analysis reveal overexpression insulin receptor various growth factor receptor tyrosine kinase chinese hamster ovary cell increase ercc-1 messenger rna (mrna) level. effect occur cell overexpressing mutate insulin receptor know impaired kinase-related signal. increase ercc-1 expression correlate resistance uv exposure. fluorescent-activated cell sorter analysis confluent cell population indicate difference cell cycle distribution. furthermore, significant relationship demonstrate relative expression ercc-1 mrna rate glucose utilization. insulin enhance accumulation ercc-1 mrna serum-deprived cell express wild-type insulin receptor. potential role activation insulin receptor related growth factor receptor ercc-1 gene expression function remains define.'], ['ass autocrine function insulin-like growth factor ii (igf-ii) balance proliferation differentiation ht29-d4 human colonic cancer cell, study expression igf-i receptor (igf-ir) insulin receptor (ir) relation state cell differentiation. igf-ir ir express undifferentiated enterocyte-like differentiate ht29-d4 cell. igf-ir two isoforms 97-kda 102-kda beta-subunit. addition, ht29-d4 cell express hybrid receptor (hr) form association two alphabeta heterodimers ir igf-ir. hr evidence 1) inhibition igf-i binding b6 anti-ir antibody 2) immunoprecipitation alpha-ir3 anti-igf-ir antibody, reveal additional 95-kda ir beta-subunit disappear heterotetrameric receptor dissociate disulfide reduction alphabeta heterodimers immunoprecipitation. like igf-ir, hr high affinity igf-i (kd, approximately 1.5 nm), bind insulin significantly latter interact native ir (kd, approximately 4 nm). differentiate ht29-d4 cell monolayer, receptor specie strongly polarize (> 97 %) toward basolateral membrane. moreover, ht29-d4 cell differentiation accompany approximately 2-fold increase number ir, whereas number igf-i-binding site unaltered. however, differentiate ht29-d4 cell, approximately 55% latter involve hr vs. approximately 20% undifferentiated ht29-d4 cell. thus, ht29-d4 cell differentiation characterize up-regulation (approximately 3-fold) level hr couple down-regulation (approximately 40 %) level native tetrameric igf-ir. alteration induced early cell differentiation process, i.e. 5 day postconfluence, remain unchanged least 21 day. take together, result suggest igf-ii autocrine loop ht29-d4 cell may trigger distinct signal pathway activates native igf-ir, predominate undifferentiated cell, activates hr, up-regulated differentiate cell.'], ['myotonic dystrophy dominantly inherit clinically variable multisystemic disorder, found cause heterozygosity trinucleotide repeat expansion mutation 3 untranslated region protein kinase gene (dm kinase). mechanism expand repeat dna result dominant biochemical defect varied clinical phenotype, know. recently propose model disease pathogenesis may occur rna level myotonic dystrophy mutant dm kinase rna expansion mutation may disrupt cellular rna metabolism general manner, evidence defect rna processing normal dm kinase gene heterozygous patient (dominant negative rna mutation). test hypothesis measure rna metabolism gene patient muscle biopsy (nine adult onset myotonic dystrophy patient, two congenital muscular dystrophy patient, four normal control, four myopathic control). focus insulin receptor gene document insulin resistance dm patient. show significant decrease insulin receptor rna total rna rna polya+ pool relative normal myopathic control muscle (p <0.002), measure relative dystrophin rna muscle sodium channel rna. also show reduction insulin receptor protein. result reinforce concept generalize rna metabolism defect myotonic dystrophy, offer possible molecular mechanism increase insulin resistance observe many myotonic dystrophy patient.']]
    tagged_corpus = read_corpus(pmids, abstracts)
    train_model(tagged_corpus, 'dummy_model')


    